makespec: Use portable path separators

bwoodsend opened the pull request:
Use forward slashes – even on Windows – in filenames in generated spec files so that the spec file is independent of the platform it was generated.

bwoodsend committed:
makespec: Use portable path separators.

Use forward slashes – even on Windows – in filenames in generated spec
files so that the spec file is independent of the platform it was
generated.

bwoodsend committed:
makespec: Use portable path separators.

Use forward slashes – even on Windows – in filenames in generated spec
files so that the spec file is independent of the platform it was
generated. (#8004)

Also fix some of the sourounding repr vs str formatting that causes
backslashes from CLI inputs to be written unescaped to the spec file.
(#9472)

rokm commented:
Hmm, I'm quite sure that the splash image path needs to receive the `repr_filepath()` treatment as well - as this is what originally sparked #8004.

rokm commented on PyInstaller/building/makespec.py:
```diff
@@ -724,32 +724,30 @@ def main(
             exe_options += "\n    icon='NONE',"
         else:
-            exe_options += "\n    icon=[%s]," % ','.join("'%s'" % escape_win_filepath(ic) for ic in icon_file)
+            exe_options += "\n    icon=[%s]," % ','.join(repr_filepath(ic) for ic in icon_file)
```
This looks suspect to me. Should it not be
```
exe_options += "\n icon=%r," % ','.join(to_posix_path(ic) for ic in icon_file)
```

instead? I intentionally replaced `repr_filepath` with non-existent `to_posix_path` because I think the `repr` part should be left out of `repr_filepath()`...

rokm replied:
I.e., we need a complete representation of the original list (and not a list of representations), so we can "load" it via execution.

rokm commented:
Is there a reason why `repr()` is included in `repr_filepath()` (i.e., why it returns representation)? I think it would make more sense for the string building part to use `%r`, while the separator-fixing-function returned only posix path.

(Actually, perhaps it would make sense for all fields in the generated template to use %r as a format string?).

bwoodsend commented:
Mind if I scrap the whole `makespec.make_variable_path()` functionality? The only substitution it knows is replacing the directory of PyInstaller itself which no one should have any reason to use?

rokm commented:
> Mind if I scrap the whole `makespec.make_variable_path()` functionality? The only substitution it knows is replacing the directory of PyInstaller itself which no one should have any reason to use?

Hmm, I have a vague recollection of trying to make sense of that part of code sometime back...

As you say, `makespec.make_variable_path` by default supports only substitution with `HOMEPATH`. And it seems to be only used by `makespec.Path` (whose name I absolutely detest), which we apply to entry-point scripts. Which have no business being rooted in PyInstaller's `HOMEPATH`...

So yeah, go ahead and scrape it.

rokm commented:
> > Mind if I scrap the whole `makespec.make_variable_path()` functionality? The only substitution it knows is replacing the directory of PyInstaller itself which no one should have any reason to use?
> 
> Hmm, I have a vague recollection of trying to make sense of that part of code sometime back...
> 
> As you say, `makespec.make_variable_path` by default supports only substitution with `HOMEPATH`. And it seems to be only used by `makespec.Path` (whose name I absolutely detest), which we apply to entry-point scripts. Which have no business being rooted in PyInstaller's `HOMEPATH`...
> 
> So yeah, go ahead and scrape it.

Ahh, `HOMEPATH` is not PyInstaller's package directory, but its parent directory (usually `site-packages`). So this was probably supposed to make spec portable if entry-point script was also in `site-packages`, "next to PyInstaller". Or maybe if for some reason you checked out `pyinstaller` repository and then placed your code into it, next to `PyInstaller` importable package. I still find it rather pointless, though.

Plus, you have a 50% chance of having it work on linux with venv, due to `venv/lib64` being symlink to `venv/lib`, and this code not resolving symlinks. So depending of which absolute path you use, it might do that replacement or not...

bwoodsend committed:
makespec: Remove *variable path* substitution.

Remove the automatic substitution of PyInstaller.HOME_PATH (most likely
the ``site-packages`` directory) in relative CLI path parameters to into
a relocatable ``os.path.join(HOME_PATH, ...)``. I can't think of any
good reason why anyone would even notice this feature exists whereas
it's more likely that a PyInstaller test would unwittingly run into it.

bwoodsend committed:
makespec: Remove *variable path* substitution.

Remove the automatic substitution of PyInstaller.HOME_PATH (most likely
the ``site-packages`` directory) in relative CLI path parameters to into
a relocatable ``os.path.join(HOME_PATH, ...)``. I can't think of any
good reason why anyone would even notice this feature exists whereas
it's more likely that a PyInstaller test would unwittingly run into it.

rokm commented on PyInstaller/building/makespec.py:
```diff
@@ -759,10 +701,7 @@ def main(
-    scripts = [make_path_spec_relative(x, specpath) for x in scripts]
-    # With absolute paths replace prefix with variable HOMEPATH.
-    scripts = list(map(Path, scripts))
+    scripts = list(map(portable_filepath, scripts))
```
I think these script paths still need to be made spec-relative, though, because the build process assumes that if they are relative, they are relative to spec directory.

For example, this used to work

`pyinstaller --clean --noconfirm program/program.py --specpath spec/`

and now it does not (and in contrast to the `HOMEPATH` thing, I imagine this is used). 

Although I suspect that this scenario is party broken anyway, because *all* given relative paths (icons, etc.) that we later treat as spec-relative would need to be made relative to spec here... But for now, we should at least maintain status quo here.

rokm commented:
Those test failures are not directly related to the PR; the matching pattern in 
https://github.com/pyinstaller/pyinstaller/blob/b2b188b5865975b67a9f79b46f2af294def0adb9/tests/functional/test_binary_vs_data_reclassification.py#L50
does not account for numbers greater than 99 (e.g., `Analysis-105.toc`). This probably depends on the test order (I recall seeing that test fail sporadically on the CI), and the changes made here might have rearranged the tests so that the issue occurs more consistently.

rokm commented on PyInstaller/building/makespec.py:
```diff
@@ -724,33 +682,31 @@ def main(
-        resources = list(map(escape_win_filepath, resources))
-        exe_options += "\n    resources=%s," % repr(resources)
+        resources = [portable_filepath(r) for r in resources]
+        exe_options += "\n    resources=%r," % repr(resources)
```
This `repr()` needs to be removed as it is now part of the format.

rokm commented on PyInstaller/building/templates.py:
```diff
@@ -94,23 +94,23 @@
-    name='%(name)s.app',
-    icon=%(icon)s,
-    bundle_identifier=%(bundle_identifier)s,
+    name=%(name)r',
```
There's a left-over trailing `'` here. And the argument should be `app_bundle_name` now, otherwise we will try to overwrite the "regular" onedir build.

rokm commented on PyInstaller/building/templates.py:
```diff
@@ -94,23 +94,23 @@
-    name='%(name)s.app',
-    icon=%(icon)s,
-    bundle_identifier=%(bundle_identifier)s,
+    name=%(name)r,
```
Also needs to use `app_bundle_name`.

bwoodsend committed:
Fix ValueError on Windows if --specpath points to a different drive

bwoodsend commented:
Brain is really not screwed on well today...

bwoodsend commented:
There we go. App bundles have their .app suffixes again.

rokm commented on PyInstaller/building/makespec.py:
```diff
@@ -710,22 +710,17 @@ def main(
         # Icon file for macOS.
-        # We need to encapsulate it into apostrofes.
-        icon_file = "'%s'" % icon_file[0]
+        icon_file = portable_filepath(icon_file[0])
```
Hmm, if `icon_file[0]` is `'NONE'`, it probably shouldn't go through `portable_filepath()` to obtain `icon_file` for bundles. Normally, it will be resolved to an absolute path and then turned back into relative (= `'NONE'`) - but if spec path happens to differ from current working directory, I suspect it will break.

bwoodsend replied:
It doesn't actually work anyway. By that point, `icon_file` is only still relevant for macOS where we don't support `'NONE'` -> no icon (it goes ahead and tries to load the image file called ./NONE).

But I suppose I can shuffle it around to be less confusing...

bwoodsend replied:
```diff
diff --git i/PyInstaller/building/makespec.py w/PyInstaller/building/makespec.py
index 97b810e65..5130735e7 100644
--- i/PyInstaller/building/makespec.py
+++ w/PyInstaller/building/makespec.py
@@ -705,18 +705,14 @@ def main(
     if uac_uiaccess:
         exe_options += "\n    uac_uiaccess=True,"
     if icon_file:
+        icon_file = [portable_filepath(ic) for ic in icon_file if ic != "NONE"]
         # Icon file for Windows.
         # On Windows, the default icon is embedded in the bootloader executable.
-        if icon_file[0] == 'NONE':
+        if not icon_file:
             exe_options += "\n    icon='NONE',"
         else:
-            exe_options += "\n    icon=%r," % [portable_filepath(ic) for ic in icon_file]
-        # Icon file for macOS.
-        icon_file = portable_filepath(icon_file[0])
-    else:
-        # On macOS, the default icon has to be copied into the .app bundle.
-        # The the value None means - use default icon.
-        icon_file = None
+            exe_options += "\n    icon=%r," % icon_file
+    icon_file = icon_file or [None]
     if contents_directory:
         exe_options += "\n    contents_directory=%r," % (contents_directory or "_internal")
     if hide_console:
@@ -841,7 +837,7 @@ def main(
         'console': console,
         'disable_windowed_traceback': disable_windowed_traceback,
         # Icon filename. Only macOS uses this item.
-        'icon': icon_file,
+        'icon': icon_file[0],
         # .app bundle identifier. Only macOS uses this item.
         'bundle_identifier': bundle_identifier,
         # argv emulation (macOS only)
```
Does that look any better to you? The optional plurality and None vs NONE stuff is more finicky than it looks.

bwoodsend replied:
Maybe do a `s/icon_file/icon_files/g` whilst I'm there?

rokm replied:
Ah, indeed. What confused me was that previously this was set to `'None'` string, but that was because it was then written to spec as a string, and became a "proper" `None`...

I'm fine with leaving it as it is then; I do not think that the updated variant really helps that all that much with the mess around the optional plurality, None vs 'NONE', etc. Although if you find it to be clearer, we can also go with as well.
